cmd/scollector/collectors/local_listener.go

```diff
@@ -0,0 +1,106 @@
+package collectors
+
+import (
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"bosun.org/cmd/scollector/conf"
+	"bosun.org/metadata"
+	"bosun.org/opentsdb"
+)
+
+func init() {
+	registerInit(func(c *conf.Conf) {
+		if c.LocalListener != "" {
+			collectors = append(collectors, &StreamCollector{F: func() <-chan *opentsdb.MultiDataPoint {
+				return c_local_listener(c.LocalListener)
+			},
+				name: fmt.Sprintf("local_listener-%s", c.LocalListener),
+			})
+		}
+	})
+}
+
+func c_local_listener(listenAddr string) <-chan *opentsdb.MultiDataPoint {
+	pm := &putMetric{}
+	pm.localMetrics = make(chan *opentsdb.MultiDataPoint, 1)
+
+	mux := http.NewServeMux()
+	mux.Handle("/api/put", pm)
+	mux.HandleFunc("/api/metadata/put", putMetadata)
+	go http.ListenAndServe(listenAddr, mux)
+
+	return pm.localMetrics
+}
+
+type putMetric struct {
+	localMetrics chan *opentsdb.MultiDataPoint
+}
+
+func (pm *putMetric) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var (
+		bodyReader io.ReadCloser
+		err        error
+	)
+
+	if r.Method != "POST" {
+		w.WriteHeader(405)
+		return
+	}
+
+	if r.Header.Get("Content-Encoding") == "gzip" {
+		if bodyReader, err = gzip.NewReader(r.Body); err != nil {
+			w.WriteHeader(500)
+			w.Write([]byte(fmt.Sprintf("Unable to decompress: %s\n", err)))
+			return
+		}
+	} else {
+		bodyReader = r.Body
+	}
+
+	if body, err := ioutil.ReadAll(bodyReader); err != nil {
```

Comment:
We need it because there's 2 ways the body can be decoded, as single data point or an array of data points. Both format are accepted by OpenTSDB. This saves us from decompressing the body 2 times.